apicast/src/oauth/keycloak.lua

```diff
@@ -0,0 +1,267 @@
+    config = keycloak_config,
+    http_client = http_client,
+    service = service
+    }, mt)
+end
+
+function _M.respond_and_exit(status, body, headers)
+  -- TODO: is there a better way to populate the response headers?..
+  if headers then
+    for name,value in pairs(headers) do
+      ngx.header[name] = value
+    end
+  end
+
+  ngx.status = status
+  ngx.print(body)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+function _M.respond_with_error(status, message)
+
+  --TODO: as per the RFC (https://tools.ietf.org/html/rfc6749#section-5.2) return WWW-Authenticate response header if 401
+  local headers = {
+    ['Content-Type'] = 'application/json;charset=UTF-8'
+  }
+  local err_msg = { error = message }
+  local body = cjson.encode(err_msg)
+  _M.respond_and_exit(status, body, headers)
+end
+
+function _M.authorize_check_params(params)
+  local response_type = params.response_type
+  local required_params = _M.params.response_type
+  if not response_type then return false, 'invalid_request' end
+  if not required_params[response_type] then return false, 'unsupported_response_type' end
+
+  for _,v in ipairs(required_params[response_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+
+  return true
+end
+
+function _M.token_check_params(params)
+  local grant_type = params.grant_type
+  local required_params = _M.params.grant_type
+  if not grant_type then return false, 'invalid_request' end
+  if not required_params[grant_type] then return false, 'unsupported_grant_type' end
+
+  for _,v in ipairs(required_params[grant_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+  return true
+end
+
+-- Parses the token - in this case we assume it's a JWT token
+-- Here we can extract authenticated user's claims or other information returned in the access_token
+-- or id_token by RH SSO
+function _M.parse_and_verify_token(self, jwt_token)
+  local jwt_obj = jwt:verify(self.config.public_key, jwt_token)
+  if not jwt_obj.verified then
+    ngx.log(ngx.INFO, "[jwt] failed verification for token: ", jwt_token)
+    return nil
+  end
+  return jwt_obj
+end
+
+function _M.check_credentials(self, params)
+  local http_client = self.http_client
+
+  if not http_client then
+    return nil, 'not initialized'
+  end
+
+  local args = {
+      app_id = params.client_id,
+      app_key = params.client_secret,
+      redirect_uri = params.redirect_uri
+    }
+  local credentials = ngx.encode_args(args)
+
+  local service = self.service
+  local service_args = ngx.encode_args({ [service.backend_authentication.type or ''] = service.backend_authentication.value, service_id = service.id })
+  local endpoint = service.backend.endpoint
+
+  if not endpoint then
+    ngx.log(ngx.WARN, 'service ', service.id, ' does not have backend endpoint configured')
+    return
+  end
+
+  local url = resty_url.join(endpoint, '/transactions/oauth_authorize.xml', "?", service_args, "&", credentials)
```

Comment:
Done in 33f7ce3. Also you can fetch the credentials from ngx.ctx introduced in 783f07b